### PR TITLE
Implement P2445R1 `forward_like()`

### LINKED ISSUE #2931: P2445R1 `forward_like()`
[P2445R1](https://wg21.link/P2445R1) `forward_like()`

Feature-test macro (expected):
```cpp
#define __cpp_lib_forward_like 202207L
```

### PR DESCRIPTION
**Pull request by frederick-vs-ja**
Fixes #2931.

Intentionally don't call `std::move`, `std::forward`, or `std::as_const`. Also move `_Can_reference` to `<utility>` for error message.

### CODE REVIEW

File: stl/inc/utility [hunk 1]
```diff
@@ -802,6 +818,27 @@ _NODISCARD constexpr underlying_type_t<_Ty> to_underlying(_Ty _Value) noexcept {
     _CSTD abort(); // likely to be called in debug mode, but can't be relied upon - already entered the UB territory
 #endif // _DEBUG
 }
+
+template <class _Ty, class _Uty>
+constexpr auto&& forward_like(_Uty&& _Ux) noexcept {
```
Comment (CaseyCarter) [inline stl/inc/utility:823]: The proposal says this should be `[[nodiscard]]`. Did you omit that for reason, or was it inadvertent?

File: stl/inc/utility [hunk 2]
```diff
@@ -802,6 +818,27 @@ _NODISCARD constexpr underlying_type_t<_Ty> to_underlying(_Ty _Value) noexcept {
     _CSTD abort(); // likely to be called in debug mode, but can't be relied upon - already entered the UB territory
 #endif // _DEBUG
 }
+
+template <class _Ty, class _Uty>
+constexpr auto&& forward_like(_Uty&& _Ux) noexcept {
+    static_assert(
+        _Can_reference<_Ty>, "The first template parameter of std::forward_like must be a referenceable type.");
```
Comment (CaseyCarter) [inline stl/inc/utility:825]:
```suggestion
        _Can_reference<_Ty>, "std::forward_like's first template argument must be a referenceable type.");
```

File: tests/std/tests/P2445R1_forward_like/test.compile.pass.cpp [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+owner_type owner{};
+constexpr owner_type& owner_lref  = owner;
+constexpr owner_type&& owner_rref = move(owner);
+
+constexpr owner_type owner_c{};
+constexpr const owner_type& owner_clref  = owner_c;
+constexpr const owner_type&& owner_crref = move(owner_c);
+
+static_assert(is_same_v<decltype(forward_like<decltype(owner)>(owner.m_mutobj)), int&&>);
```
Comment (CaseyCarter) [inline tests/std/tests/P2445R1_forward_like/test.compile.pass.cpp:30]: I'm not fond of the `decltype` usage in these `static_assert`s, I have a mild preference to see the type spelled out. With blocks of six I admit it makes little difference, so let's so what other reviewers have to say.

**Review by CaseyCarter — CHANGES_REQUESTED** [review]

File: stl/inc/utility (hunk 1 above)
Comment (frederick-vs-ja) [reply stl/inc/utility]: I'm sorry for just forgetting it. Perhaps because [`[[nodiscard]]` was absent in the first post on cppreference](https://en.cppreference.com/mwiki/index.php?title=std::forward_like&oldid=141395) of which I wasn't aware. I'm fixing this now.

File: tests/std/tests/P2445R1_forward_like/test.compile.pass.cpp (hunk 3 above)
Comment (frederick-vs-ja) [reply tests/std/tests/P2445R1_forward_like/test.compile.pass.cpp]: I've also avoided nested `decltype` now. I don't know whether this is wanted.

File: tests/std/tests/P2445R1_forward_like/test.compile.pass.cpp [hunk 4]
```diff
@@ -27,52 +27,94 @@ constexpr owner_type owner_c{};
...
-static_assert(is_same_v<decltype(forward_like<decltype(owner_clref)>(owner_clref.m_crref)), const int&>);
-
-static_assert(is_same_v<decltype(forward_like<decltype(owner_crref)>(owner_crref.m_mutobj)), const int&&>);
-static_assert(is_same_v<decltype(forward_like<decltype(owner_crref)>(owner_crref.m_mutlref)), const int&&>);
-static_assert(is_same_v<decltype(forward_like<decltype(owner_crref)>(owner_crref.m_mutrref)), const int&&>);
-
-static_assert(is_same_v<decltype(forward_like<decltype(owner_crref)>(owner_crref.m_cobj)), const int&&>);
-static_assert(is_same_v<decltype(forward_like<decltype(owner_crref)>(owner_crref.m_clref)), const int&&>);
-static_assert(is_same_v<decltype(forward_like<decltype(owner_crref)>(owner_crref.m_crref)), const int&&>);
+using owner_t              = decltype(owner);
```
Comment (CaseyCarter) [inline tests/std/tests/P2445R1_forward_like/test.compile.pass.cpp:30]: I have some concerns about test coverage. The expressions `owner`, `owner_lref`, and `owner_rref` are all lvalues of type `owner_type`. Given an lvalue `x` of type `owner_type`, the expressions `x.m_mutobj`, `x.m_mutlref`, and `x.m_mutrref` are all lvalues of type `int`. We pass an lvalue of type `int` to `forward_like<owner_t>`, `forward_like<owner_lref_t>`, and `forward_like<owner_rref_t>` three times each. This seems like overkill when once each would suffice. We similarly pass lvalues of type `const int` three times each to `forward_like<owner_t>`, `forward_like<owner_lref_t>`, and `forward_like<owner_rref_t>`.

For the `const` paths, the numbers are a little different. The expressions `owner_c`, `owner_clref`, and `owner_crref` are all lvalues of type `const owner_type`. Given an lvalue `x` of type `const owner_type`, the expressions `x.m_mutobj`, `x.m_cobj`, `x.m_clref` and `x.m_crref` are all lvalues of type `const int`. The expressions `x.m_mutlref` and `x.m_mutrref` are lvalues of type `int`. So instead of 3 `const` and 3 non-`const` lvalue `int`s per set of calls we have 4 `const` and 2 non-`const` lvalue `int`s per set of calls.

I think the `owner` class and the complicated naming is all obfuscating what's happening here and the fact that we are missing coverage despite having quite a bit of redundant coverage. I think we could greatly simplify all of this to:
```c+
struct T {}; // class type so const prvalues are valid
using CT = const T;
using U  = int;
using CU = const U;

T t{};
const T& ct = t;

static_assert(is_same_v<decltype(forward_like<U>(T{})), T&&>);
static_assert(is_same_v<decltype(forward_like<U>(CT{})), CT&&>);
static_assert(is_same_v<decltype(forward_like<U>(t)), T&&>);
static_assert(is_same_v<decltype(forward_like<U>(ct)), CT&&>);
static_assert(is_same_v<decltype(forward_like<U>(move(t))), T&&>);
static_assert(is_same_v<decltype(forward_like<U>(move(ct))), CT&&>);

static_assert(is_same_v<decltype(forward_like<CU>(T{})), CT&&>);
static_assert(is_same_v<decltype(forward_like<CU>(CT{})), CT&&>);
static_assert(is_same_v<decltype(forward_like<CU>(t)), CT&&>);
static_assert(is_same_v<decltype(forward_like<CU>(ct)), CT&&>);
static_assert(is_same_v<decltype(forward_like<CU>(move(t))), CT&&>);
static_assert(is_same_v<decltype(forward_like<CU>(move(ct))), CT&&>);

static_assert(is_same_v<decltype(forward_like<U&>(T{})), T&>);
static_assert(is_same_v<decltype(forward_like<U&>(CT{})), CT&>);
static_assert(is_same_v<decltype(forward_like<U&>(t)), T&>);
static_assert(is_same_v<decltype(forward_like<U&>(ct)), CT&>);
static_assert(is_same_v<decltype(forward_like<U&>(move(t))), T&>);
static_assert(is_same_v<decltype(forward_like<U&>(move(ct))), CT&>);

static_assert(is_same_v<decltype(forward_like<CU&>(T{})), CT&>);
static_assert(is_same_v<decltype(forward_like<CU&>(CT{})), CT&>);
static_assert(is_same_v<decltype(forward_like<CU&>(t)), CT&>);
static_assert(is_same_v<decltype(forward_like<CU&>(ct)), CT&>);
static_assert(is_same_v<decltype(forward_like<CU&>(move(t))), CT&>);
static_assert(is_same_v<decltype(forward_like<CU&>(move(ct))), CT&>);

static_assert(is_same_v<decltype(forward_like<U&&>(T{})), T&&>);
static_assert(is_same_v<decltype(forward_like<U&&>(CT{})), CT&&>);
static_assert(is_same_v<decltype(forward_like<U&&>(t)), T&&>);
static_assert(is_same_v<decltype(forward_like<U&&>(ct)), CT&&>);
static_assert(is_same_v<decltype(forward_like<U&&>(move(t))), T&&>);
static_assert(is_same_v<decltype(forward_like<U&&>(move(ct))), CT&&>);

static_assert(is_same_v<decltype(forward_like<CU&&>(T{})), CT&&>);
static_assert(is_same_v<decltype(forward_like<CU&&>(CT{})), CT&&>);
static_assert(is_same_v<decltype(forward_like<CU&&>(t)), CT&&>);
static_assert(is_same_v<decltype(forward_like<CU&&>(ct)), CT&&>);
static_assert(is_same_v<decltype(forward_like<CU&&>(move(t))), CT&&>);
static_assert(is_same_v<decltype(forward_like<CU&&>(move(ct))), CT&&>);
```
`|{const, non-const} x {prvalue, lvalue, xvalue}| == 6`, so we have precisely 6 sets of 6 test cases each to get full coverage. Does that make sense?

**Review by CaseyCarter — CHANGES_REQUESTED** [review]

File: tests/std/tests/P2445R1_forward_like/test.compile.pass.cpp (hunk 4 above)
Comment (frederick-vs-ja) [reply tests/std/tests/P2445R1_forward_like/test.compile.pass.cpp]: Thank you very much! I used to be considering "owner and member" in the paper, and be afraid that `int` is not suitable as an "owner". But I can go with it now because `int` and `T` are capable to emulate the "far" state.

I'm taking your test coverage, making the comments a bit wordy but more clear to me.

**Review by CaseyCarter — APPROVED** [review]
LGTM!

File: stl/inc/utility [hunk 5]
```diff
@@ -802,6 +818,26 @@ _NODISCARD constexpr underlying_type_t<_Ty> to_underlying(_Ty _Value) noexcept {
     _CSTD abort(); // likely to be called in debug mode, but can't be relied upon - already entered the UB territory
 #endif // _DEBUG
 }
+
+template <class _Ty, class _Uty>
+_NODISCARD constexpr auto&& forward_like(_Uty&& _Ux) noexcept {
+    static_assert(_Can_reference<_Ty>, "std::forward_like's first template argument must be a referenceable type.");
+
+    using _UnrefU = remove_reference_t<_Uty>;
```
Comment (strega-nil-ms) [inline stl/inc/utility:826]: To make this closer to the wording of the standard, adding `using _UnrefT = remove_reference_t<_Ty>` here would be good - we do a `remove_reference_t` in both cases anyways.

Comment (strega-nil-ms) [inline stl/inc/utility:826]: To make this closer to the wording of the standard, adding `using _UnrefT = remove_reference_t<_Ty>` here would be good - we do a `remove_reference_t` in both cases anyways.

File: stl/inc/utility [hunk 6]
```diff
@@ -802,6 +818,26 @@ _NODISCARD constexpr underlying_type_t<_Ty> to_underlying(_Ty _Value) noexcept {
     _CSTD abort(); // likely to be called in debug mode, but can't be relied upon - already entered the UB territory
 #endif // _DEBUG
 }
+
+template <class _Ty, class _Uty>
+_NODISCARD constexpr auto&& forward_like(_Uty&& _Ux) noexcept {
+    static_assert(_Can_reference<_Ty>, "std::forward_like's first template argument must be a referenceable type.");
+
+    using _UnrefU = remove_reference_t<_Uty>;
+    if constexpr (is_lvalue_reference_v<_Ty>) {
```
Comment (strega-nil-ms) [inline stl/inc/utility:827]: I believe it's a more obvious copying of the standard to put the `is_const_v` on the outside of the if block.

**Review by strega-nil-ms — COMMENTED** [review]
I don't really have a "problem" with how it's written right now, but I _would_ prefer it if this was written closer to the wording of the paper; something like:

```cxx
using _UnrefT = remove_reference_t<_Ty>;
using _UnrefU = remove_reference_t<_Uty>;
if constexpr (is_const_v<_UnrefT>) {
    // _Copy_const = const _UnrefU
    if constexpr (is_rvalue_reference_v<_Ty&&>) {
        return static_cast<const _UnrefU&&>(_Ux);
    } else {
        return static_cast<const _UnrefU&>(_Ux);
    }
} else {
    // _Copy_const = _UnrefU
    if constexpr (is_rvalue_reference_v<_Ty&&>) {
        return static_cast<_UnrefU&&>(_Ux);
    } else {
        return static_cast<_UnrefU&>(_Ux);
    }
}
```

File: tests/std/tests/P2445R1_forward_like/test.compile.pass.cpp [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+
+struct T {}; // class type so const-qualification is not stripped from a prvalue
+using CT = const T;
+using U  = int;
+using CU = const U;
+
+T t{};
+const T& ct = t;
+
+static_assert(is_same_v<decltype(forward_like<U>(T{})), T&&>);
```
Comment (strega-nil-ms) [inline tests/std/tests/P2445R1_forward_like/test.compile.pass.cpp:17]: It would be nice here to swap `U` and `T`, so that `U` corresponds to `_Uty` and `T` corresponds to `_Ty`.

File: stl/inc/utility (hunk 5 above)
Comment (frederick-vs-ja) [reply stl/inc/utility:827]: Most review comments are addressed. But I prefer to leave the hard error caused by invalid `_Ty&&` in `static_assert`, so that we can control the error message.

File: tests/std/tests/P2445R1_forward_like/test.compile.pass.cpp (hunk 7 above)
Comment (frederick-vs-ja) [reply tests/std/tests/P2445R1_forward_like/test.compile.pass.cpp]: I missed renaming of  `t` and `ct`. Will do later.

File: tests/std/tests/P2445R1_forward_like/test.compile.pass.cpp [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+static_assert(is_same_v<decltype(forward_like<T&&>(move(cu))), CU&&>);
+
+static_assert(is_same_v<decltype(forward_like<CT&&>(U{})), CU&&>);
+static_assert(is_same_v<decltype(forward_like<CT&&>(CU{})), CU&&>);
+static_assert(is_same_v<decltype(forward_like<CT&&>(u)), CU&&>);
+static_assert(is_same_v<decltype(forward_like<CT&&>(cu)), CU&&>);
+static_assert(is_same_v<decltype(forward_like<CT&&>(move(u))), CU&&>);
+static_assert(is_same_v<decltype(forward_like<CT&&>(move(cu))), CU&&>);
+
+int main() {} // COMPILE-ONLY
```
Comment (StephanTLavavej) [inline tests/std/tests/P2445R1_forward_like/test.compile.pass.cpp:59]: These tests are good, and verify the trickiest part of the implementation. Additionally, it would be good to quickly check:

* That the function is `constexpr`
* That the function is `noexcept`
* That the function returns a reference to its argument

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
@frederick-vs-ja @CaseyCarter @strega-nil-ms Thanks, this looks good! I added a preprocessor comment and slightly expanded the test coverage (mostly out of principle, not any real concern about library or compiler bugs).

**Review by CaseyCarter — APPROVED** [review]

File: stl/inc/utility (hunk 5 above)
Comment (strega-nil-ms) [reply stl/inc/utility:827]: Oh, to be clear I wasn't attempting to ask you to remove that, sorry!

**Review by strega-nil-ms — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

**Comment by StephanTLavavej** [comment]
Thanks for implementing this feature - I *like* to see the STL moving *forward*! :joy_cat: :heart_eyes_cat: :rocket:
